lib/propolis/src/enlightenment/hyperv/mod.rs

```diff
@@ -0,0 +1,313 @@
+
+        // TLFS section 3.13 says that the hypercall page "becomes disabled" if
+        // the guest OS ID register is cleared after the hypercall register is
+        // set. It also specifies that attempts to set the Enabled bit in that
+        // register will be ignored if the guest OS ID is zeroed, so handle this
+        // case by clearing the hypercall MSR's Enabled bit but otherwise
+        // leaving the hypercall page untouched (as would happen if the guest
+        // manually cleared this bit).
+        if value == 0 {
+            info!(&self.log, "guest cleared HV_X64_MSR_GUEST_OS_ID");
+            inner.msr_hypercall_value.clear_enabled();
+        } else {
+            info!(
+                &self.log,
+                "guest set HV_X64_MSR_GUEST_OS_ID";
+                "value" => value
+            );
+        }
+
+        inner.msr_guest_os_id_value = value;
+        WrmsrOutcome::Handled
+    }
+
+    /// Handles a write to the HV_X64_MSR_HYPERCALL register. See TLFS section
+    /// 3.13.
+    fn handle_wrmsr_hypercall(&self, value: u64) -> WrmsrOutcome {
+        let mut inner = self.inner.lock().unwrap();
+        let old = inner.msr_hypercall_value;
+
+        // Per the spec, this MSR is immutable once the Locked bit is set.
+        if old.locked() {
+            return WrmsrOutcome::Handled;
+        }
+
+        // If this MSR is written when no guest OS ID is set, the Enabled bit is
+        // cleared and the write succeeds.
+        let mut new = MsrHypercallValue(value);
+        if inner.msr_guest_os_id_value == 0 {
+            new.clear_enabled();
+        }
+
+        // If the Enabled bit is set, expose the hypercall instruction sequence
+        // at the requested GPA. The TLFS specifies that this raises #GP if the
+        // selected physical address is outside the bounds of the guest's
+        // physical address space.
+        //
+        // The TLFS describes this page as an "overlay" that "covers whatever
+        // else is mapped to the GPA range." The spec is ambiguous as to whether
+        // this means that the page's previous contents must be restored if the
+        // page is disabled or its address later changes. Empirically, most
+        // guest OSes don't appear to move the page after creating it,
+        // and at least some other hypervisors don't bother with saving and
+        // restoring the old page's contents. So, for simplicity, simply write
+        // the hypercall instruction sequence to the requested page and call it
+        // a day.
+        let outcome = if new.enabled() {
+            let memctx = self
+                .acc_mem
+                .access()
+                .expect("guest memory is always accessible during wrmsr");
+
+            let region = GuestRegion(new.gpa(), PAGE_SIZE);
+            if let Some(mapping) = memctx.writable_region(&region) {
+                write_overlay_page(&mapping, &hypercall_page_contents());
```

Comment:
FWIW, a minor incompatibility here: the TLFS also says that once a page is donated this way, writes to it should produce a #GP, but I'm not sure bhyve gives us the ability to rewrite the protection masks for an individual page in a previously-allocated guest memory region. But I'm not inclined to lose much sleep over it, since the guest will need to map this page as executable, and if it's wise it will hopefully not map it as WX...